flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogOptions.java

```diff
@@ -57,4 +87,18 @@
                             "Number of threads to use to perform cleanup in case an upload is discarded "
                                     + "(and not cleaned up by JM). "
                                     + "If the cleanup doesn't keep up then task might be back-pressured.");
+
+    public static final ConfigOption<Integer> NUM_UPLOAD_THREADS =
+            ConfigOptions.key("dstl.dfs.upload.num-threads")
+                    .intType()
+                    .defaultValue(5)
+                    .withDescription("Number of threads to use for upload.");
```

Comment:
I think the assumption that async checkpoint part uploads are **always** using the same shared pool is incorrect. For example RocksDB uses separate threads to upload, and the IO executor is only used to wait for the result. 

The advantages are
- better isolation (e.g. heavy network traffic from the upstream does not affect snapshot uploading => checkpointing)
- fine-grained tuning is possible
- in some scenarios it's easier to debug (i.e. to see where changelog uploader is stuck, not why IO thread isn't uploading)

The disadvantages are flip-sides of the latter two: the additional config option and in more difficult to debug in other cases .

Additionally, here **scheduled** executor is needed, while current IO executor is just `Executor`.

Reply:
> This sounds like RocksDB issue 

I guess it was a design decision to (easily) throttle the number of request to DFS (S3). This is yet another reason for a separate thread pool here.

> I can see two reasons against using the same thread pool: ...2. having actions with different priority

It's not only about the priority, but also about isolation, as I wrote above. If there is an issue with uploading (e.g. bug or request throttling) then it will consume all IO threads - regardless of the priority. This makes it harder to debug because it will manifest itself in many different ways, giving false signals.

> I don't believe anyone will want to bother how much he should bump individual thread pools

I'd say users shouldn't bother at all and the defaults should be enough. But once there **is** some issue it might be helpful to see what are threads in each pool doing - and maybe adjust accordingly.

>  Besides I don't see this being an issue in this case either?

I looked at the code, and it seems can block releasing partitions. Maybe there are or will be some other effects?

> In that case it should be solved probably in a better way by still sharing thread pool, but with more sophisticated actions scheduler

I think it's better only because of the other point - ease of configuration. Otherwise, separate thread pools are simpler in code and more robust.

Besides that, async checkpointing phase (regardless of the backend) currently uses its own pool created in `StreamTask`, not the common IO pool.